### PR TITLE
xorg/system: Modernize

### PR DESCRIPTION
**Pull request by jwillikers**
Use tools.system.package_manager ang tools.gnu.PkgConfig from the `conan` namespace.
This greatly reduces complexity of the package.

Specify library name and version:  **xorg/system**
---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by ericLemanissier** [comment]
1. This fails on FreeBSD with a lot of `ld: error: unable to find library -lxcb` when linking the test package cf https://github.com/bincrafters/system-packages-checks/runs/7356617446?check_suite_focus=true#step:3:17496. These libraries have properly been installed above. Is the library path propagated ?
2. This fails on debian 10 with `dpkg-query: no packages found matching libxcb-util-dev` and `dpkg-query: no packages found matching libxcb-util0-dev` https://github.com/bincrafters/system-packages-checks/runs/7356633944?check_suite_focus=true#step:6:814
3. alma linux fails with [`ERROR: while executing system_requirements(): Error 100 while executing yum check-update -y`](https://github.com/bincrafters/system-packages-checks/runs/7356632831?check_suite_focus=true#step:6:214) This has to be solved in conan
4. [fedora](https://github.com/bincrafters/system-packages-checks/runs/7356633486?check_suite_focus=true#step:6:171), [opensuse](https://github.com/bincrafters/system-packages-checks/runs/7356633403?check_suite_focus=true#step:6:171) and [centos](https://github.com/bincrafters/system-packages-checks/runs/7356633558?check_suite_focus=true#step:6:177) don't install any package. You have to add DNF handling to this PR

File: recipes/xorg/all/conanfile.py [hunk 1]
```diff
@@ -12,81 +14,49 @@ class ConanXOrg(ConanFile):
...
+        apt.install(["libx11-dev", "libx11-xcb-dev", "libfontenc-dev", "libice-dev", "libsm-dev", "libxau-dev", "libxaw7-dev",
+                     "libxcomposite-dev", "libxcursor-dev", "libxdamage-dev", "libxdmcp-dev", "libxext-dev", "libxfixes-dev",
+                     "libxi-dev", "libxinerama-dev", "libxkbfile-dev", "libxmu-dev", "libxmuu-dev",
+                     "libxpm-dev", "libxrandr-dev", "libxrender-dev", "libxres-dev", "libxss-dev", "libxt-dev", "libxtst-dev",
+                     "libxv-dev", "libxvmc-dev", "libxxf86vm-dev", "xtrans-dev", "libxcb-render0-dev",
+                     "libxcb-render-util0-dev", "libxcb-xkb-dev", "libxcb-icccm4-dev", "libxcb-image0-dev",
+                     "libxcb-keysyms1-dev", "libxcb-randr0-dev", "libxcb-shape0-dev", "libxcb-sync-dev", "libxcb-xfixes0-dev",
+                     "libxcb-xinerama0-dev", "xkb-data", "libxcb-dri3-dev", "uuid-dev"], update=True, check=True)
+        apt.install_substitutes(
+            ["libxcb-util-dev", "libxcb-util0-dev"], update=True, check=True)
```
Comment (ericLemanissier) [inline recipes/xorg/all/conanfile.py:35]:
```suggestion
            [["libxcb-util-dev"], ["libxcb-util0-dev"]], update=True, check=True)
```

File: recipes/xorg/all/conanfile.py [hunk 2]
```diff
@@ -12,81 +14,49 @@ class ConanXOrg(ConanFile):
...
-        self.cpp_info.components[name].set_property("component_version", pkg_config.version[0])
-        self.cpp_info.components[name].set_property(
-            "pkg_config_custom_content",
-            "\n".join("%s=%s" % (key, value) for key,value in variables.items()))
+        Yum(self).install(["libxcb-devel", "libfontenc-devel", "libXaw-devel", "libXcomposite-devel",
+                           "libXcursor-devel", "libXdmcp-devel", "libXtst-devel", "libXinerama-devel",
+                           "libxkbfile-devel", "libXrandr-devel", "libXres-devel", "libXScrnSaver-devel", "libXvMC-devel",
+                           "xorg-x11-xtrans-devel", "xcb-util-wm-devel", "xcb-util-image-devel", "xcb-util-keysyms-devel",
+                           "xcb-util-renderutil-devel", "libXdamage-devel", "libXxf86vm-devel", "libXv-devel",
+                           "xcb-util-devel", "libuuid-devel", "xkeyboard-config-devel"], update=True, check=True)
```
Comment (ericLemanissier) [inline recipes/xorg/all/conanfile.py:43]:
```suggestion
        Dnf(self).install(["libxcb-devel", "libfontenc-devel", "libXaw-devel", "libXcomposite-devel",
                           "libXcursor-devel", "libXdmcp-devel", "libXtst-devel", "libXinerama-devel",
                           "libxkbfile-devel", "libXrandr-devel", "libXres-devel", "libXScrnSaver-devel", "libXvMC-devel",
                           "xorg-x11-xtrans-devel", "xcb-util-wm-devel", "xcb-util-image-devel", "xcb-util-keysyms-devel",
                           "xcb-util-renderutil-devel", "libXdamage-devel", "libXxf86vm-devel", "libXv-devel",
                           "xcb-util-devel", "libuuid-devel", "xkeyboard-config-devel"], update=True, check=True)

```

File: recipes/xorg/all/conanfile.py [hunk 3]
```diff
@@ -1,7 +1,9 @@
-from conans import ConanFile, tools
-from conans.errors import ConanException, ConanInvalidConfiguration
+from conan import ConanFile
+from conan.tools.gnu import PkgConfig
+from conan.tools.system.package_manager import Apt, Yum, PacMan, Pkg, Zypper
```
Comment (ericLemanissier) [inline recipes/xorg/all/conanfile.py:3]:
```suggestion
from conan.tools.system.package_manager import Apt, Yum, PacMan, Pkg, Zypper, Dnf
```

File: recipes/xorg/all/conanfile.py (hunk 1 above)
Comment (jwillikers) [reply recipes/xorg/all/conanfile.py]: Done.

File: recipes/xorg/all/conanfile.py (hunk 2 above)
Comment (jwillikers) [reply recipes/xorg/all/conanfile.py:43]: Done.

File: recipes/xorg/all/conanfile.py (hunk 3 above)
Comment (jwillikers) [reply recipes/xorg/all/conanfile.py]: Done. Thanks for all the help!

**Comment by ericLemanissier** [comment]
please, don't force-push, it prevents from doing incremental reviews

**Comment by jwillikers** [comment]
> please, don't force-push, it prevents from doing incremental reviews

Sorry about that, I wasn't aware.

File: recipes/xorg/all/conanfile.py [hunk 4]
```diff
@@ -12,81 +14,56 @@ class ConanXOrg(ConanFile):
...
+        apt.install(["libx11-dev", "libx11-xcb-dev", "libfontenc-dev", "libice-dev", "libsm-dev", "libxau-dev", "libxaw7-dev",
+                     "libxcomposite-dev", "libxcursor-dev", "libxdamage-dev", "libxdmcp-dev", "libxext-dev", "libxfixes-dev",
+                     "libxi-dev", "libxinerama-dev", "libxkbfile-dev", "libxmu-dev", "libxmuu-dev",
+                     "libxpm-dev", "libxrandr-dev", "libxrender-dev", "libxres-dev", "libxss-dev", "libxt-dev", "libxtst-dev",
+                     "libxv-dev", "libxvmc-dev", "libxxf86vm-dev", "xtrans-dev", "libxcb-render0-dev",
+                     "libxcb-render-util0-dev", "libxcb-xkb-dev", "libxcb-icccm4-dev", "libxcb-image0-dev",
+                     "libxcb-keysyms1-dev", "libxcb-randr0-dev", "libxcb-shape0-dev", "libxcb-sync-dev", "libxcb-xfixes0-dev",
+                     "libxcb-xinerama0-dev", "xkb-data", "libxcb-dri3-dev", "uuid-dev"], update=True, check=True)
+        apt.install_substitutes(
+            [["libxcb-util-dev"], ["libxcb-util0-dev"]], update=True, check=True)
```
Comment (ericLemanissier) [inline recipes/xorg/all/conanfile.py:35]:
```suggestion
            ["libxcb-util-dev"], ["libxcb-util0-dev"], update=True, check=True)
```

Comment (jwillikers) [reply recipes/xorg/all/conanfile.py]: Done.

**Comment by ericLemanissier** [comment]
status update:

1. This fails on FreeBSD with a lot of ld: error: unable to find library -lxcb when linking the test package cf https://github.com/bincrafters/system-packages-checks/runs/7368977364?check_suite_focus=true#step:3:17487. These libraries have properly been installed above. Is the library path propagated ? This fails at link time, not at run time, so `LD_LIBRARY_PATH` is not relevant here
2. [alma linux](https://github.com/bincrafters/system-packages-checks/runs/7368984232?check_suite_focus=true#step:6:214), [centos](https://github.com/bincrafters/system-packages-checks/runs/7368984251?check_suite_focus=true#step:6:184) and [fedora ](https://github.com/bincrafters/system-packages-checks/runs/7368984210?check_suite_focus=true#step:6:206) fail with `ERROR: while executing system_requirements(): Error 100 while executing yum check-update -y`. This has to be solved in conan, because it just signals that there are some update available. This error code should be ignored (it was ignored in the previous system package handler)
3. [opensuse](https://github.com/bincrafters/system-packages-checks/runs/7368984701?check_suite_focus=true#step:6:150) does not install any package, even if your PR properly calls Zypper. I have no idea what's wrong for this one
4. raspbery fails, it will be fixed with https://github.com/conan-io/conan/pull/11654

**Comment by jgsogo** [comment]
> @jgsogo @danimtb is there a plan to add `tools.system.package_manager:mode=install` config on C3I ?

I guess this is the equivalent in `v2` to `conan config set general.sysrequires_mode=enabled`, right?

**Comment by ericLemanissier** [comment]
yes : https://docs.conan.io/en/latest/reference/conanfile/tools/system/package_manager.html#configuration-properties-that-affect-how-system-package-managers-are-invoked
The difference is that it can be set in:

- global.conf
- the profile
- the commande line with `-c`

**Comment by jwillikers** [comment]
@ericLemanissier The issue on FreeBSD should probably be fixed in Conan, though I've implemented a bit of a hack to workaround it. The crux of the problem is that specifying the libraries as `system` libraries requires that the libraries be found in the default search locations, i.e. `/usr/lib` and `/usr/lib64` because `-l` is passed only the name of the library. Link directories are not actually communicated to the compiler when using the `CMakeDeps` generator. Instead, I think that Conan uses the full path of the library when linking against it. That means you either get the full path to the library for normal libraries or just the library name itself when it is a __system__ library.

Considering that FreeBSD installs packages with Pkg to `/usr/local/` and `/usr/local/lib` is not in the default search path, either Conan must automatically account for this or FreeBSD users need to add configuration for this situation by setting the `tools.build:exelinkflags` and `tools.build:sharedlinkflags` to include `-L/usr/local/lib`.

Ironically, Conan _configures include directories for system libraries_ using the `INTERFACE_INCLUDE_DIRECTORIES` property, so finding the includes is not a problem.

**Comment by ericLemanissier** [comment]
Thanks for your research

> Link directories are not actually communicated to the compiler when using the `CMakeDeps` generator

This is the underlying issue right ? Is it reported ?

**Comment by SSE4** [comment]
@ericLemanissier why compiler should care about link directories? I believe only linker needs em.

**Comment by ericLemanissier** [comment]
I made a leap, and assumed the directories are not communicated to the linker either. Is it what you found out @jwillikers ?
That would explain why FreeBSD does not find xorg system libs => https://github.com/bincrafters/system-packages-checks/runs/7368977364?check_suite_focus=true#step:3:17487

**Comment by jwillikers** [comment]
> Thanks for your research
> 
> > Link directories are not actually communicated to the compiler when using the `CMakeDeps` generator
> 
> This is the underlying issue right ? Is it reported ?

I believe this is actually an intentional design decision since it was brought up in issue conan-io/conan#10837. I believe the solution there which enables a CMakeDeps configuration value in the package is only applied by Conan when on Windows. Since libraries packaged via Conan are typically linked against by their full path when using the `CMakeDeps` generator, this isn't typically a problem. It's probably worth opening in issue for this specific use case, especially since the `Pkg` package manager tool in Conan is much less effective when `/usr/local/lib` isn't included in the system library search directories.

**Comment by jwillikers** [comment]
> I made a leap, and assumed the directories are not communicated to the linker either. Is it what you found out @jwillikers ? That would explain why FreeBSD does not find xorg system libs => https://github.com/bincrafters/system-packages-checks/runs/7368977364?check_suite_focus=true#step:3:17487

Yes, that's exactly what I meant.

**Comment by ericLemanissier** [comment]
I confirm that you fixed FreeBSD with this change : https://github.com/bincrafters/system-packages-checks/runs/7453182952?check_suite_focus=true#step:3:17549

**Comment by jwillikers** [comment]
I've opened an issue upstream for how Conan should handle system libraries installed to `/usr/local/lib`, which is done by all *BSD's: conan-io/conan#11688`.

**Comment by ericLemanissier** [comment]
status update:

1. [alma linux](https://github.com/bincrafters/system-packages-checks/runs/7368984232?check_suite_focus=true#step:6:214), [centos](https://github.com/bincrafters/system-packages-checks/runs/7368984251?check_suite_focus=true#step:6:184) and [fedora ](https://github.com/bincrafters/system-packages-checks/runs/7368984210?check_suite_focus=true#step:6:206) fail with `ERROR: while executing system_requirements(): Error 100 while executing yum check-update -y`. This has to be solved in conan, because it just signals that there are some update available. This error code should be ignored (it was ignored in the previous system package handler). I reported it https://github.com/conan-io/conan/issues/11661, it will be solved in conan 1.51.0
2. [opensuse](https://github.com/bincrafters/system-packages-checks/runs/7368984701?check_suite_focus=true#step:6:150) does not install any package, even if your PR properly calls Zypper. This should be fixed by https://github.com/conan-io/conan/pull/11660
3. raspbery fails, it has been fixed with https://github.com/conan-io/conan/pull/11654, in conan 1.51.0

All these should be fixed with conan 1.51.0, so the only blocking point remaining is the setting of `tools.system.package_manager:mode=install` config on C3I

**Comment by ericLemanissier** [comment]
@jwillikers it seems it should work now https://github.com/conan-io/conan-center-index/pull/11859#pullrequestreview-1047735377
Can you close/reopen ?

**Comment by jwillikers** [comment]
> @jwillikers it seems it should work now https://github.com/conan-io/conan-center-index/pull/11859/files#diff-77f023b99d3d58008351d3e82fc06e6d06ba1bc2da9e41be6329b7fa4f419f05R10 Can you close/reopen ?

Sorry, had to rebase and force push. Nothing has actually changed in the recipe.

**Review by ericLemanissier — DISMISSED** [review]

**Review by danimtb — APPROVED** [review]

**Review by ericLemanissier — APPROVED** [review]

**Pull request merged**